### PR TITLE
[C++] Support ObjectRef arguments in cluster mode

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?

Support ObjectRef arguments on cluster mode, also handled reference counting. This PR don't support nested ObjectRef, it will support in another PR.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 1]
```diff
@@ -115,6 +115,35 @@ std::vector<bool> AbstractRayRuntime::Wait(const std::vector<std::string> &ids,
...
 }
 
+rpc::Address GetOwnershipInfo(const std::string &object_id_str) {
+  if (ConfigInternal::Instance().run_mode == RunMode::SINGLE_PROCESS) {
+    return {};
+  }
+  ray::internal::outer_object_id = object_id_str;
+  auto object_id = ray::ObjectID::FromBinary(object_id_str);
+  rpc::Address address;
+  std::string serialized_object_status;
```
Comment (kfstorm) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:125]: Could you also leverage `serialized_object_status`? We're trying to support this in Java. See https://github.com/ray-project/ray/pull/19984

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 2]
```diff
@@ -115,6 +115,35 @@ std::vector<bool> AbstractRayRuntime::Wait(const std::vector<std::string> &ids,
   return object_store_->Wait(StringIDsToObjectIDs(ids), num_objects, timeout_ms);
 }
 
+rpc::Address GetOwnershipInfo(const std::string &object_id_str) {
+  if (ConfigInternal::Instance().run_mode == RunMode::SINGLE_PROCESS) {
+    return {};
+  }
+  ray::internal::outer_object_id = object_id_str;
```
Comment (kfstorm) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:122]: I believe the thread-local outer object id should be set at the beginning of deserializing an object and reset at the end of deserializing an object, not in `GetOwnershipInfo`.

See https://github.com/ray-project/ray/blob/5ffa0bb01f89bdd1678f3e795a7b6a8bf0e616e2/java/runtime/src/main/java/io/ray/runtime/object/ObjectStore.java#L125-L129

File: cpp/src/ray/runtime/abstract_ray_runtime.h [hunk 3]
```diff
@@ -118,5 +118,10 @@ class AbstractRayRuntime : public RayRuntime {
   void Execute(const TaskSpecification &task_spec);
   PlacementGroup GeneratePlacementGroup(const std::string &str);
 };
+
+rpc::Address GetOwnershipInfo(const std::string &object_id);
+void RegisterOwnershipInfoAndResolveFuture(const std::string &object_id,
+                                           const std::string &outer_object_id,
+                                           const rpc::Address &owner_addr);
```
Comment (kfstorm) [inline cpp/src/ray/runtime/abstract_ray_runtime.h:125]: Why aren't they member functions of `AbstractRayRuntime`?

File: cpp/src/ray/runtime/task/task_executor.cc [hunk 4]
```diff
@@ -150,6 +150,8 @@ Status TaskExecutor::ExecuteTask(
     msgpack::sbuffer sbuf;
 
     if (is_ref_arg) {
+      RegisterOwnershipInfoAndResolveFuture(
+          ref.object_id(), ray::internal::outer_object_id, ref.owner_address());
```
Comment (kfstorm) [inline cpp/src/ray/runtime/task/task_executor.cc:154]: `RegisterOwnershipInfoAndResolveFuture` should be invoked every time deserializing an ObjectRef, not just during deserializing task arguments, but also during `ray::Get`. If you invoke it only when deserializing task arguments, how do you plan to support arguments of type `std::vector<ray::ObjectRef>`?

File: cpp/src/ray/test/cluster/cluster_mode_test.cc [hunk 5]
```diff
@@ -408,6 +405,17 @@ TEST(RayClusterModeTest, TaskWithPlacementGroup) {
...
+  auto obj = ray::Put(std::string("aaa"));
+  auto r = ray::Task(GetVal).Remote(obj);
+  EXPECT_EQ(*r.Get(), "aaa");
+
+  auto obj1 = ray::Put(1);
+  auto obj2 = ray::Put(2);
+  auto r1 = ray::Task(Add).Remote(obj1, obj2);
+  EXPECT_EQ(*r1.Get(), 3);
+}
+
```
Comment (kfstorm) [inline cpp/src/ray/test/cluster/cluster_mode_test.cc:418]: Please add test cases about `ray::Get(std::vector<ray::ObjectRef>)` and `ray::Task` with a `std::vector<ray::ObjectRef>` argument.

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 6]
```diff
@@ -135,14 +135,19 @@ std::vector<std::unique_ptr<::ray::TaskArg>> TransformArgs(
...
+          AbstractRayRuntime::GetInstance()->GetOwnershipInfoInternal(*arg.id),
+          /*call_site=*/"");
+
+      // auto id = ObjectID::FromBinary(*arg.id);
+      // auto owner_address = ray::rpc::Address{};
+      // if (ConfigInternal::Instance().run_mode == RunMode::CLUSTER) {
+      //   auto &core_worker = CoreWorkerProcess::GetCoreWorker();
+      //   owner_address = core_worker.GetOwnerAddress(id);
+      // }
+      // ray_arg = absl::make_unique<ray::TaskArgByReference>(id, owner_address,
```
Comment (MissiontoMars) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:149]: Should be removed.

File: cpp/src/ray/runtime/abstract_ray_runtime.cc [hunk 7]
```diff
@@ -343,5 +348,52 @@ PlacementGroup AbstractRayRuntime::GetPlacementGroup(const std::string &name) {
...
+    return {};
+  }
+
+  auto object_id = ray::ObjectID::FromBinary(object_id_str);
+  return CoreWorkerProcess::GetCoreWorker().GetOwnerAddress(object_id);
+  // rpc::Address address;
+  // std::string serialized_object_status;
+  // CoreWorkerProcess::GetCoreWorker().GetOwnershipInfo(object_id, &address,
+  //                                                     &serialized_object_status);
+  // return address;
```
Comment (MissiontoMars) [inline cpp/src/ray/runtime/abstract_ray_runtime.cc:368]: Should be removed.

File: cpp/src/ray/runtime/abstract_ray_runtime.h [hunk 8]
```diff
@@ -104,6 +104,14 @@ class AbstractRayRuntime : public RayRuntime {
 
   virtual std::vector<PlacementGroup> GetAllPlacementGroups();
   virtual PlacementGroup GetPlacementGroupById(const std::string &id);
+  virtual std::string GetOwnershipInfo(const std::string &object_id);
+  rpc::Address GetOwnershipInfoInternal(const std::string &object_id);
+  virtual void RegisterOwnershipInfoAndResolveFuture(const std::string &object_id_str,
+                                                     const std::string &outer_object_id,
+                                                     const std::string &owner_addr);
+  void RegisterOwnershipInfoAndResolveFutureInternal(const std::string &object_id,
```
Comment (MissiontoMars) [inline cpp/src/ray/runtime/abstract_ray_runtime.h:112]: `RegisterOwnershipInfoAndResolveFutureInternal` should be private or protected.

File: cpp/include/ray/api/function_manager.h [hunk 9]
```diff
@@ -51,11 +51,12 @@ msgpack::sbuffer PackError(std::string error_msg);
 using ArgsBuffer = msgpack::sbuffer;
 using ArgsBufferList = std::vector<ArgsBuffer>;
 
-using RemoteFunction = std::function<msgpack::sbuffer(const ArgsBufferList &)>;
+using RemoteFunction = std::function<msgpack::sbuffer(const ArgsBufferList &,
```
Comment (MissiontoMars) [inline cpp/include/ray/api/function_manager.h:54]: Actually I think we can add more comments for cpp worker.

File: cpp/include/ray/api/function_manager.h [hunk 10]
```diff
@@ -111,11 +116,12 @@ struct Invoker {
 
  private:
   template <typename T>
-  static inline T ParseArg(const ArgsBuffer &args_buffer, bool &is_ok) {
+  static inline T ParseArg(const ArgsBuffer &args_buffer, bool &is_ok,
+                           const std::string &object_id) {
```
Comment (MissiontoMars) [inline cpp/include/ray/api/function_manager.h:120]: Better to use different `ParseArg` implementation based on T.

**Comment by bveeramani** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've switched our code formatter from YAPF to Black (see #21311).

To prevent issues with merging your code, here's what you'll need to do:

1. Install Black
```
pip install -I black==21.12b0
```

2. Format changed files with Black
```
curl -o format-changed.sh https://gist.githubusercontent.com/bveeramani/42ef0e9e387b755a8a735b084af976f2/raw/7631276790765d555c423b8db2b679fd957b984a/format-changed.sh
chmod +x ./format-changed.sh
./format-changed.sh
rm format-changed.sh
```

3. Commit your changes.
```
git add --all
git commit -m "Format Python code with Black"
```

4. Merge master into your branch.
```
git pull upstream master
```

5. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated `format.sh`.

**Comment by kfstorm** [comment]
## :bangbang: ACTION REQUIRED :bangbang:

We've updated our formatting configuration for C++ code. (see #22725)

This PR includes C++ code change. To prevent issues with merging your code, here's what you'll need to do:

1. Merge the latest changes from upstream/master branch into your branch.

```bash
git pull upstream master
git merge upstream/master
```

2. Resolve merge conflicts (if necessary).

After running these steps, you'll have the updated C++ formatting configuration.

3. Format changed files.

```bash
scripts/format.sh
```

4. Commit your changes.

```bash
git add --all
git commit -m "Format C++ code"
```

**Pull request closed without merging**